### PR TITLE
Cost 1835 handle default partition data

### PR DESCRIPTION
**Pull request by Red-HAP**
## Tickets

* **Jira:** [COST-1835](https://issues.redhat.com/browse/COST-1835)
* **[Sentry](https://sentry.io/organizations/project-koku/issues/2618461980/events/latest/?project=5183023)**

## Description

PR #3153 should be merged before this.

This is a fix in two parts:
1. Create an updated suite of partition get_or_create functions that will:
    1. Check to see if a default partition is present and if not, create one
    2. Check to see if an existing default partition has data that will overlap with the requested partition
    3. If data overlap exists, then move that data from the default partition to the requested partition
2. Change the trino processing to utilize the main function call to execute the above logic
3. Get or create partitions for requested-partition - 1 partition-period; requested-partition; requested-partition + 1 partition-period. This is to handle cases of providers that can send information in a billing period that may overflow a partition boundary.

## Testing

This requires Parquet/Trino processing.

Feed some files through that overlap a partition boundary. After processing, there should be no data in the default partition.
There should be partitions that contain the data and _probably_ 1 extra partition that contains no data (depending on if the overflow is one or both partition bounds).

### Console Output

[Output](https://ci.ext.devshift.net/job/project-koku-koku-pr-check/1435/console)

### Summary

```
16:44:43 =========================== short test summary info ============================
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-last-30-days]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-last-month-daily]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-two-months-ago-daily]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-last-month]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-two-months-ago]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-date-min-today]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-node-date-45ago-today]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-last-30-days]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-last-month-daily]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-two-months-ago-daily]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-last-month]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-two-months-ago]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-date-min-today]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_compute_reports.py::test_api_ocp_compute_validate_total[usage-project-date-45ago-today]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_memory_reports.py::test_api_ocp_memory_validate_total[usage-project-date-min-today]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_memory_reports.py::test_api_ocp_memory_validate_total[usage-node-project-date-min-today]
16:44:43 FAILED lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_gcp_instance_reports.py::test_api_gcp_instance_report_filter_match_date_range[two-months-ago-daily]
16:44:43 ERROR lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test__ocp_cost_reports.py::test_api_ocp_cost_group_by_exclusive_matching[project]
16:44:43 ERROR lib64/python3.6/site-packages/iqe_cost_management/tests/rest_api/v1/test_source.py::test_api_azure_source_returns_cost_model
16:44:43 = 17 failed, 5592 passed, 172 skipped, 848 deselected, 26 warnings, 2 errors in 2343.41s (0:39:03) =
```

### CODE REVIEW

File: koku/koku/pg_partition.py [hunk 1]
```diff
@@ -1668,3 +1679,193 @@ def repartition_default_data(schema_name=None, partitioned_table_name=None):
...
+        cur.execute(chk_sql)
+        res = cur.fetchone()
+
+    if res[0]:
+        LOG.info(f"Overlapping data found in the default partition for {chk_where}")
+        restore = part_rec["partition_parameters"].copy()
+        if default_partition.partition_type == default_partition.RANGE:
+            pp_from = ciso8601.parse_datetime(part_rec["partition_parameters"]["from"]).date()
+            pp_to = ciso8601.parse_datetime(part_rec["partition_parameters"]["to"]).date()
+            delta = relativedelta(years=random.randrange(100, 7000))
```
Comment (maskarb) [inline koku/koku/pg_partition.py:1756]: what is this for?

Comment (Red-HAP) [reply koku/koku/pg_partition.py:1756]: @maskarb 
I was concerned that, if I used a constant, there might be a collision from concurrent processes.
I confess I might have been overthinking this, though.

**Review by adberglund — APPROVED** [review]

**Pull request merged**